test/cpp/api/misc.cpp

```diff
@@ -41,4 +41,17 @@ TEST_F(AutogradTest, CanTakeDerivativesOfZeroDimTensors) {
 TEST_F(AutogradTest, CanPassCustomGradientInputs) {
   z.sum().backward(torch::ones({}) * 2);
   ASSERT_TRUE(x.grad().allclose(y * 2));
-}
\ No newline at end of file
+}
+
+TEST_F(AutogradTest, CannotModifyVersionInVarTypeMode) {
+  y.add_(1);
+  ASSERT_THROWS_WITH(z.sum().backward(), "modified by an inplace operation");
+}
+
+TEST_F(AutogradTest, CannotModifyVersionInNonVarTypeMode) {
+  {
+    at::AutoNonVariableTypeMode non_var_type_mode(true);
+    y.add_(1);  // This uses non-Variable type dispatch
+  }
+  ASSERT_THROWS_WITH(z.sum().backward(), "modified by an inplace operation");
+}
```

Comment:
@gchanan These two are the use cases we care about, and they will still work after Variable/Tensor merge. Detailed discussion is at the end of the PR description.